// mjml-accordion/lib/AccordionText.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _mjmlCore = require("mjml-core");

var MjAccordionText = /*#__PURE__*/function (_BodyComponent) {
  (0, _inherits2["default"])(MjAccordionText, _BodyComponent);

  function MjAccordionText() {
    (0, _classCallCheck2["default"])(this, MjAccordionText);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(MjAccordionText).apply(this, arguments));
  }

  (0, _createClass2["default"])(MjAccordionText, [{
    key: "getStyles",
    value: function getStyles() {
      return {
        td: {
          background: this.getAttribute('background-color'),
          'font-size': this.getAttribute('font-size'),
          'font-family': this.getAttribute('font-family'),
          color: this.getAttribute('color'),
          'padding-bottom': this.getAttribute('padding-bottom'),
          'padding-left': this.getAttribute('padding-left'),
          'padding-right': this.getAttribute('padding-right'),
          'padding-top': this.getAttribute('padding-top'),
          padding: this.getAttribute('padding')
        },
        table: {
          width: '100%',
          'border-bottom': this.getAttribute('border')
        }
      };
    }
  }, {
    key: "renderContent",
    value: function renderContent() {
      return "\n      <td\n        ".concat(this.htmlAttributes({
        "class": this.getAttribute('css-class'),
        style: 'td'
      }), "\n      >\n        ").concat(this.getContent(), "\n      </td>\n    ");
    }
  }, {
    key: "render",
    value: function render() {
      return "\n      <div\n        ".concat(this.htmlAttributes({
        "class": 'mj-accordion-content'
      }), "\n      >\n        <table\n          ").concat(this.htmlAttributes({
        'cell-spacing': '0',
        'cell-padding': '0',
        style: 'table'
      }), "\n        >\n          <tbody>\n            <tr>\n              ").concat(this.renderContent(), "\n            </tr>\n          </tbody>\n        </table>\n      </div>\n    ");
    }
  }]);
  return MjAccordionText;
}(_mjmlCore.BodyComponent);

exports["default"] = MjAccordionText;
(0, _defineProperty2["default"])(MjAccordionText, "endingTag", true);
(0, _defineProperty2["default"])(MjAccordionText, "allowedAttributes", {
  'background-color': 'color',
  'font-size': 'unit(px)',
  'font-family': 'string',
  color: 'color',
  'padding-bottom': 'unit(px,%)',
  'padding-left': 'unit(px,%)',
  'padding-right': 'unit(px,%)',
  'padding-top': 'unit(px,%)',
  padding: 'unit(px,%){1,4}'
});
(0, _defineProperty2["default"])(MjAccordionText, "defaultAttributes", {
  'font-size': '13px',
  padding: '16px'
});
module.exports = exports.default;
