// mjml-spacer/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _mjmlCore = require("mjml-core");

var _conditionalTag = _interopRequireDefault(require("mjml-core/lib/helpers/conditionalTag"));

var MjSpacer = /*#__PURE__*/function (_BodyComponent) {
  (0, _inherits2["default"])(MjSpacer, _BodyComponent);

  function MjSpacer() {
    (0, _classCallCheck2["default"])(this, MjSpacer);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(MjSpacer).apply(this, arguments));
  }

  (0, _createClass2["default"])(MjSpacer, [{
    key: "getStyles",
    value: function getStyles() {
      return {
        div: {
          height: this.getAttribute('height')
        }
      };
    }
  }, {
    key: "render",
    value: function render() {
      var height = this.getAttribute('height');
      return "\n      ".concat((0, _conditionalTag["default"])("\n        <table role=\"presentation\" border=\"0\" cellpadding=\"0\" cellspacing=\"0\"><tr><td height=\"".concat(parseInt(height, 10), "\" style=\"vertical-align:top;height:").concat(height, ";\">\n      ")), "\n      <div\n        ").concat(this.htmlAttributes({
        style: 'div'
      }), "\n      >\n        &nbsp;\n      </div>\n      ").concat((0, _conditionalTag["default"])("\n        </td></tr></table>\n      "), "\n    ");
    }
  }]);
  return MjSpacer;
}(_mjmlCore.BodyComponent);

exports["default"] = MjSpacer;
(0, _defineProperty2["default"])(MjSpacer, "allowedAttributes", {
  border: 'string',
  'border-bottom': 'string',
  'border-left': 'string',
  'border-right': 'string',
  'border-top': 'string',
  'container-background-color': 'color',
  'padding-bottom': 'unit(px,%)',
  'padding-left': 'unit(px,%)',
  'padding-right': 'unit(px,%)',
  'padding-top': 'unit(px,%)',
  padding: 'unit(px,%){1,4}',
  'vertical-align': 'enum(top,bottom,middle)',
  width: 'unit(px,%)',
  height: 'unit(px,%)'
});
(0, _defineProperty2["default"])(MjSpacer, "defaultAttributes", {
  height: '20px'
});
module.exports = exports.default;
